src/microbe_stage/editor/PatchMapDrawer.cs

```diff
@@ -1109,10 +1116,67 @@ private void AddPatchNode(Patch patch, Vector2 position)
 
         patch.ApplyPatchEventVisuals(node);
 
+        var playerSpecies = patch.FindSpeciesByID(PlayerSpeciesID);
+        if (playerSpecies != null)
+        {
+            AddPlayerPopulationIndicators(patch, playerSpecies, node, position);
+        }
+
         patchNodeContainer.AddChild(node);
         nodes.Add(node.Patch, node);
     }
 
+    private void AddPlayerPopulationIndicators(Patch patch, Species playerSpecies, Control node, Vector2 position)
+    {
+        var playerPopulationIndicatorAmount = (int)Math.Ceiling(patch.GetSpeciesSimulationPopulation(playerSpecies) *
+            Constants.PLAYER_POPULATION_POPULATION_FOR_PER_INDICATOR);
+
+        var indicatorExcess = Math.Clamp(playerSpeciesPopulationIndicators.Count - playerPopulationIndicatorAmount,
+            0,
+            playerSpeciesPopulationIndicators.Count);
+
+        // Hide excess from the end of the list
+        for (int i = 0; i < indicatorExcess; ++i)
```

Comment:
If you want to keep things this way, I'll accept it, but I think a slightly easier way to program this would be to use a variable like, `nextIndicatorIndex` that starts at 0 and goes up for each used indicator. Then the excess indicators can be hidden by looping from the ending nextIndicatorIndex value up to the indicators count. And the `noCached` variable can be calculated as `nextIndicatorIndex >= playerSpeciesPopulationIndicators.Count`.

Reply:
Now with the persistent index for used indicators, I think this hiding excess indicators code needs to be put into a single place that is called *after* the entire map is drawn. Otherwise the last indicators get hidden over and over (and I think your indicatorExcess math was not updated anyway so that would lead to a bug in any case).